reconcile/dashdotdb_dora.py

```diff
@@ -0,0 +1,480 @@
+        _, ref_to = self.get_repo_ref_for_sha(saastarget, promotion_sha)
+
+        return (saastarget, deployment, RepoChanges(repo_url, ref_from, ref_to))
+
+    def get_repo_ref_for_sha(
+        self, saastarget: SaasTarget, sha: str
+    ) -> tuple[Optional[str], Optional[str]]:
+        saas_file_yaml = self.app_interface_get_file(saastarget.path, ref=sha).decode()
+        saas_file = yaml.safe_load(saas_file_yaml)
+
+        for rt in saas_file["resourceTemplates"]:
+            if saastarget.resource_template != rt["name"]:
+                continue
+
+            for target in rt["targets"]:
+                if saastarget.namespace == target["namespace"]["$ref"]:
+                    return (rt["url"], target["ref"])
+
+        return (None, None)
+
+    def get_dep_db_deployments(
+        self, app_env: AppEnv, since: datetime
+    ) -> list[Deployment]:
+        LOG.info(
+            "Fetching deployments from DeploymentDB for %s since %s",
+            app_env,
+            since,
+        )
+        return self.dep_db.deployments(app_env, since)
+
+    def compare(self, rc: RepoChanges) -> tuple[RepoChanges, list[Commit]]:
+        if not rc.repo_url:
+            return rc, []
+
+        LOG.info("Fetching commits %s", rc)
+        if rc.repo_url.startswith("https://github.com"):
+            commits = self._github_compare_commits(rc)
+        elif rc.repo_url.startswith("https://gitlab.cee.redhat.com"):
+            commits = self._gitlab_compare_commits(rc)
+        else:
+            raise Exception(f"Unknown git hosting {rc.repo_url}")
+
+        return rc, commits
+
+    def get_latest(self, app_env: AppEnv) -> Optional[datetime]:
+        endpoint = f"{self.dashdotdb_url}/api/v1/dora/latest"
+        response = self._do_get(
+            endpoint,
+            {
+                "app_name": app_env.app_name,
+                "env_name": app_env.env_name,
+            },
+        )
+
+        if response.status_code == 404:
+            return None
+
+        return datetime.fromisoformat(response.json()["finish_timestamp"])
+
+    def _gitlab_compare_commits(self, rc: RepoChanges) -> list[Commit]:
+        if not rc.repo_url:
+            return []
+
+        prefix = "https://gitlab.cee.redhat.com/"
+        repo = rc.repo_url[rc.repo_url.startswith(prefix) and len(prefix) :]
+
+        project = self.gl.projects.get(repo)
+        result = project.repository_compare(rc.ref_from, rc.ref_to)
+
+        return [
+            Commit(
+                rc.repo_url,
+                commit["id"],
+                datetime.fromisoformat(commit["committed_date"]),
+            )
+            for commit in result["commits"]
+        ]
+
+    def _github_compare_commits(self, rc: RepoChanges) -> list[Commit]:
+        if not rc.repo_url:
+            return []
+
+        prefix = "https://github.com/"
+        repo = rc.repo_url[rc.repo_url.startswith(prefix) and len(prefix) :]
+        gh = GithubRepositoryApi(repo, self.gh_token)
```

Comment:
Note that the `repo` argument is variable. I would need to build a dict per `repo`. Is it worth it considering this integration runs on a cron?

Reply:
It's true that the initialization is not passive. I will fix this.

Reply:
@jmelis it seems we can reuse the same `Github` object to fetch different repos, `GithubRepositoryApi` just for single repo usage, if we design `GithubRepositoryApi` to accept `Github` object instead of creating it inside `__init__`, then it's very easy to support your use case here. Or you can simply init one `Github` instance, use it to fetch repos.

Reply:
Is it OK if I add a note in the code and not solve this? I don't care about performance that much, and the memory usage will not skyrocket.

Reply:
sure, need to note the runtime of this integration is a cronjob, resource cleanup is not done via code, but the termination of the process. I may include this as part of github cleanup once [pygithub 2.0.2rc1](https://github.com/PyGithub/PyGithub/milestone/44) released.

Reply:
I added the info as a class documentation. Resolving this. Please reopen if you disagree.